src/Compilers/CSharp/Portable/Symbols/TypeUnification.cs

```diff
@@ -37,6 +40,72 @@ public static bool CanUnify(TypeSymbol t1, TypeSymbol t2)
             return result;
         }
 
+        /// <summary>
+        /// Determines a substitution of type parameters on <paramref name="extension"/>
+        /// that yields <paramref name="type"/>.
+        /// The substitution should not touch any of the containing type's type parameters
+        /// and it should substitute all of the extension's own type parameters.
+        /// </summary>
+        public static bool CanImplicitlyExtend(NamedTypeSymbol extension, TypeSymbol type, out AbstractTypeParameterMap? map)
+        {
+            Debug.Assert(extension is not null);
+            Debug.Assert(type is not null);
+
+            var extensionUnderlyingType = extension.ExtendedTypeNoUseSiteDiagnostics;
+            Debug.Assert(extensionUnderlyingType is not null);
+
+            // PROTOTYPE we'll want to adjust the handling for differences that aren't relevant to the CLR, such as object/dynamic
+            if (TypeSymbol.Equals(extensionUnderlyingType, type, TypeCompareKind.CLRSignatureCompareOptions))
+            {
+                map = null;
+                return true;
+            }
+
+            MutableTypeMap? substitution = null;
+            bool result = CanUnifyHelper(extensionUnderlyingType, type, onlySubstituteInLHS: true, ref substitution);
+#if DEBUG
+            if (result && (extensionUnderlyingType is not null && type is not null))
+            {
+                var substitutedUnderlyingType = SubstituteAllTypeParameters(substitution, TypeWithAnnotations.Create(extensionUnderlyingType));
+                Debug.Assert(substitutedUnderlyingType.Type.Equals(type, TypeCompareKind.CLRSignatureCompareOptions));
+            }
+#endif
+
+            // In error scenarios where we end up with unsubstituted type parameters,
+            // we reject the extension.
+            foreach (var typeParameter in extension.TypeParameters)
+            {
+                if (substitution is null || !substitution.Contains(typeParameter))
+                {
+                    map = null;
+                    return false;
+                }
+            }
+
+            // We cannot allow any of the type parameters of the extension's containing type to be substituted
+            if (hasSubstitutionForContainingTypeTypeParameter(extension.ContainingType, substitution))
+            {
+                map = null;
+                return false;
+            }
+
+            map = substitution;
+            return result;
+
+            static bool hasSubstitutionForContainingTypeTypeParameter(NamedTypeSymbol? containingType, MutableTypeMap? substitution)
+            {
+                if (containingType is null || substitution is null)
+                {
+                    return false;
+                }
+
+                TypeSymbol? foundTypeParameter = containingType.VisitType(
```

Comment:
>containingType.VisitType [](http://example.com/codeflow?start=49&length=24)

I think this is overkill, and while, after some deep thinking, I can convince myself that all type parameters that we can run into while visiting the containing type are container's type parameters, I think a more direct and clear implementation would be preferred. What I have in mind specifically, is a loop through containing types checking its ```TypeParameters```.

Reply:
Thanks. That's simpler indeed